### PR TITLE
Bring back slice::ref_slice as slice::from_ref.

### PR DESCRIPTION
**Pull request by whitequark**
These functions were deprecated and removed in 1.5, but such simple
functionality shouldn't require using unsafe code, and it isn't
cluttering libstd too much.

The original removal was quite contentious (see #27774), since then
we've had precedent for including such nuggets of functionality (see rust-lang/rfcs#1789),
and @nikomatsakis has provided a lot of use cases in https://github.com/rust-lang/rfcs/pull/1789#issuecomment-314640034.
Hence this PR.

I'm not too sure what to do with stability, feel free to correct me.
It seems pointless to go through stabilization for these functions though.

cc @aturon

### CODE REVIEW

**Comment by whitequark** [comment]
I didn't do anything to bring back `Option::as_slice` and `Result::as_slice` because `x.map_or(&[], ref_slice)` is almost as short and just as easy to understand.

**Comment by stjepang** [comment]
Maybe it's just me, but, looking at the name alone, it's hard to tell what `slice::ref_slice` actually does.
Perhaps `slice::from_ref` would sound better? Or `slice::single_ref`? 🚲

**Comment by sfackler** [comment]
I'm not particularly opposed to bringing this back, but I do agree with @stjepang's naming concern. We may want to land these as unstable initially, though I don't really know what good that would do in this specific case.

@rfcbot fcp merge

**Comment by Mark-Simulacrum** [comment]
I've added the waiting-on-crater label for now so that if we decide to go ahead with this we do a crater run -- I think it's worth doing that just in case others have implemented ref_slice (which could play into the naming bikeshed, too, so probably best after fcp).

**Comment by sfackler** [comment]
These are free functions, so I don't think there's as much of a concern with collisions as you'd have with new methods.

**Comment by scottmcm** [comment]
Since the naming is uncertain and insta-stable is proposed anyway, could they just be
```rust
impl<'a, T> From<&'a T> for &'a [T] { ... }
impl<'a, T> From<&'a mut T> for &'a mut [T] { ... }
```

**Comment by whitequark** [comment]
Personally, I think `slice::from_ref` is an obviously better name than `slice::ref_slice`. If I thought about naming at all (I did not, I just copied them from old rustc sources) then I would have chosen that too.

A `From` impl seems more disruptive to me.

**Comment by alexcrichton** [comment]
I'd personally prefer to initially add these as unstable (but perhaps push on stabilization soon after landing), and I agree with @whitequark that a name like `from_ref` may be a bit nicer!

**Comment by Ixrec** [comment]
Also strongly agree with `from_ref`. My first guess as to what `slice::ref_slice` would do was take a slice of Ts and convert it to a slice of refs of T.

**Comment by whitequark** [comment]
Already renamed it, just not reflected in the PR.

**Comment by carols10cents** [comment]
ping @BurntSushi waiting for your ticky box [here](https://github.com/rust-lang/rust/pull/45306#issuecomment-337450327)!

**Comment by steveklabnik** [comment]
My crate that provides these functions also provides an `Option` variant: https://github.com/steveklabnik/ref_slice/blob/master/src/lib.rs

I'm not sure if they're worth bringing in too, figured I'd bring it up.

**Comment by whitequark** [comment]
@steveklabnik https://github.com/rust-lang/rust/pull/45306#issuecomment-336711884

**Comment by alexcrichton** [comment]
I'm going to go ahead and check @BurntSushi's checkbox as this is just unstable anyway.

File: src/libcore/slice/mod.rs [hunk 1]
```diff
@@ -2450,6 +2450,22 @@ pub unsafe fn from_raw_parts_mut<'a, T>(p: *mut T, len: usize) -> &'a mut [T] {
     mem::transmute(Repr { data: p, len: len })
 }
 
+/// Converts a reference to T into a slice of length 1 (without copying).
+#[stable(feature = "from_ref", since = "1.22.0")]
```
Comment (kennytm) [inline src/libcore/slice/mod.rs]: I don't think we intend to insta-stable these?

**Review by kennytm — CHANGES_REQUESTED** [review]

**Comment by kennytm** [comment]
@bors r-

If the final FCP checkbox expects "this is just unstable anyway" the current code certainly isn't unstable.

**Comment by alexcrichton** [comment]
Oops sorry, thanks for catching that @kennytm!

**Comment by whitequark** [comment]
Can you please guide me on the stability markers here?

**Comment by kennytm** [comment]
@whitequark 

1. File a [tracking issue](https://forge.rust-lang.org/feature-guide.html#tracking-issue)
2. Replace those `#[stable]` attributes with

      ```rust
      #[unstable(feature = "from_ref", issue = "45678")]
      ```

3. Write an unstable-book entry in `src/doc/unstable-book/src/library-features/from-ref.md` if you like.

**Comment by whitequark** [comment]
Tracking issue filed as https://github.com/rust-lang/rust/issues/45703, no unstable book entry since it seems far too trivial.

**Comment by whitequark** [comment]
Stability annotation updated.

**Review by kennytm — APPROVED** [review]

**Comment by kennytm** [comment]
Thanks! It should be ready to go after fixing the build failure.

```
[00:02:34] error: use of unstable library feature 'from_ref' (see issue #45703)
[00:02:34]    --> /checkout/src/liballoc/slice.rs:123:23
[00:02:34]     |
[00:02:34] 123 | pub use core::slice::{from_ref, from_ref_mut};
[00:02:34]     |                       ^^^^^^^^
[00:02:34]     |
[00:02:34]     = help: add #![feature(from_ref)] to the crate attributes to enable
[00:02:34] 
[00:02:34] error: use of unstable library feature 'from_ref' (see issue #45703)
[00:02:34]    --> /checkout/src/liballoc/slice.rs:123:33
[00:02:34]     |
[00:02:34] 123 | pub use core::slice::{from_ref, from_ref_mut};
[00:02:34]     |                                 ^^^^^^^^^^^^
[00:02:34]     |
[00:02:34]     = help: add #![feature(from_ref)] to the crate attributes to enable
```

**Comment by whitequark** [comment]
Updated.

**Comment by steveklabnik** [comment]
Library stuff doesn't need an unstable book entry as the api docs are docs;
the unstable book pages get generated automatically.

**Pull request merged**